### ISSUE TITLE
Question about distributed learning

### ISSUE THREAD

**Issue by aTan-aka-Xellos** [gcp/leela-zero#1]
Hello! 
Since it required huge amount of computing resources to teach network - is it possible to create some distributed system, where everyone who is willing can join and contribute their machine resources? 
According their paper, 40 days is needed to overcome previous version, but for AlphaGo Lee version 3 days is enough. So it seems less than 150 years is needed :D 
If 10-100 people contribute machine resources - it seems possible to make some progress. Also may donation is option to rent some powerful GPU based server (this should be even  simpler since no need to build any distributed system).
Sorry if question is silly :) 

Could you estimate - does it possible to achieve any tangible progress with this network learning on GTX1080? (by tangible I mean - to achieve at least low dan lvl for a month of time for example)

Thank you!

**Comment by gcp** [comment]
Yes, we are working on the server portion of the distributed system. It is probably possible to get very decent performance with 10-100 people in a few months, or less. The idea is also to start with an even smaller network to see if the system works correctly, and where we end up strength-wise. I will post more info here and on the computer-go mailinglist when we have something ready to test.

Aside from the server you need a small self-play GTP client (I have this ready and will upload it to a separate github repo soon) and a script to fetch the network weights from the server and upload the games back. The code in this repo also needs 2 minimal tweaks to include more randomness in the games.

Low dan level should be achievable, I think. You only need a small network for that (~6 residual layers or so). The supervised network that is available as an example probably already reaches ~1 dan level on a GTX 1080, and it was only trained for 2 days on a single GPU. (Of course, I cheated by not bootstrapping from self-play)

Despite that bad news that replicating the final results will *literally* take ages, the good news is that initial progress goes very quickly.

**Comment by aTan-aka-Xellos** [comment]
@gcp thank you for fast answer! 
Could you also clarify - do you plan to provide compiled build? Since not everyone familiar with C++ :) Or I missed link somewhere?

**Comment by gcp** [comment]
Yes, sure. I wonder if there's some way to do it automatically on github.

**Comment by Alpha3031** [comment]
You can distribute binaries using [Github Releases](https://help.github.com/articles/creating-releases/). Not exactly automatic, but I think that's the easiest way. I wonder how many people will be willing to donate compute time compared to something Fishtest.

**Comment by gafferongames** [comment]
Why distribute the training when Google provides access to TPU based processors through their cloud? You have access to the same hardware that Google used to train AlphaGo... just start a kickstarter or a Patreon to cover the cost of training...

**Comment by apetresc** [comment]
@gafferongames [Cloud TPUs](https://cloud.google.com/tpu/) are still in "preview" mode, i.e they reach out to you if you get to have access, but as far as I know nobody who applied through the public sign-up sheet has actually gotten in yet. Nor have they given any details about pricing of Cloud TPU, beyond the 1,000 free ones they're going to be donating to certain academic groups.

**Comment by gcp** [comment]
Various reasons:

1) The costs ramp up rather quickly. DeepMind's CEO estimated that they used about $25 million in computing resources. I think people are more comfortable lending their machine than they are lending their money. In any case I have zero interest in asking for people's money.\
2) I can't even get a GCP account. It's not available to European individuals (only businesses), or my account is wedged in an indeterminate state and it's impossible to get help or get it fixed. (I understand such experiences with Google Customer Support are not rare, and a reason not to depend on Google infrastructure) Amazon EC2 GPU instances aren't very cost effective and relatively slow.
3) My understanding is that using the TPUs requires using TensorFlow. I'd rather have a program that can run with minimal setup on anyone's machine. Anyone's free to write a TensorFlow version of this, run 1000's of instances on the Google Cloud at their own cost, and submit the data.

**Comment by gcp** [comment]
Small update here: the missing randomization parts are implemented. I managed to build a Windows binary (probably required if we're hoping many people join!).

Still needed: Finishing up the self-play tool, and then beta-testing the server setup.

I am testing different training configurations with a human database (supervised learning) to get some idea for the parameters (learning rate, decay, value vs policy weighting), the performance on normal GPUs, and to be able to optimize the search parameters (UCT coeff etc) in leela-zero.

**Comment by aTan-aka-Xellos** [comment]
https://github.com/yenw/computer-go-dataset could it be useful as additional database? Since it contains thouthands of FineArt/DeepZen/CGI games with pro and top amateurs?

**Comment by gcp** [comment]
Only for calibrating with a supervised learning dataset. I already have a database of comparable size, but thanks for the link. (AlphaGo Zero's paper used a smaller dataset of only KGS games, so the size shouldn't be an issue)

**Comment by xelxebar** [comment]
I'm jumping in the middle of a conversation here, but would it be feasible to write a BOINC application to do the distribution for us?

BOINC is essentially a portal of distributed computation projects and makes it easy to contribute computer cycles to whatever you find personally interesting. It houses projects like SETI@Home, CERN's physics stuff, prime number searches, etc. I think a lot of non-Go players would even be interested in contributing.

**Comment by FlorinAndrei** [comment]
Once distributed training is ready to go, I'll donate some time on my Titan X.

What might help would be binary releases once in a while. Win64, whatever's the current latest Ubuntu LTS, etc. This will attract a wider base of users.

**Comment by Ka-zam** [comment]
I'll donate time on my GTX1080 for this.

**Comment by yoonws** [comment]
I will also donate time on my GTX1080. I am student of Go in Korea, My dream is professional Go player.
I really like board game and Leela.

I have a pure curious about distributed learning system. If distribute learning is possible, there are many students of Go in Korea who want to participate.

Is it possible to learn network with seperated computer without real time network connection? 

For example, if we run learning algorithm on single computer it may build small network.
If we adjust that networks, it may just reduce biased error where error come from choice of training set.

How can it is possible.

**Comment by gcp** [comment]
>Is it possible to learn network with seperated computer without real time network connection?

In theory, yes, as long as you can download a new network periodically and upload the data periodically. Depending on the delay in updating, you'll have a slower learning-feedback loop, though, so it's not ideal. And the client would have to be written to deal with this, which certainly isn't going to be true for the first versions.

>If we adjust that networks

What would "adjust" even mean here.

**Comment by yoonws** [comment]
What I was meaning is merge,
I thought we devide total move for distribute learning.
For example,  moves1-20, moves 10-20.

The complicity of Go is come from uncertainty for large move. We can't expect for large number of move even though we have learned a lot.

But with well trained value network, story is different. We know who is good, and who is bad without total play game.

We can decision who will win at anytime, with well trained value network.
That means we can devide total number of move for distributed learning.

If we look them local moves (1-20, 10-20) it became easy game. 

That is why I thought we are doing adjust. I thought we merge small networks. 

I am not familiar with programming. Sorry for amateur like reply.

**Comment by gcp** [comment]
Sorry, but I completely fail to understand what you are proposing.

The whole point is to learn an accurate value network. If we already had one then most of the effort wouldn't be necessary. If the score is very much to one side, then the program can resign and terminate the game early (but there are caveats here - see the original paper which addresses them well).

Even so I don't understand what you would want to achieve by only looking at 10 moves from a total game, or how this would help distribution for clients that aren't connected to the internet, or anything really.

**Comment by gcp** [comment]
There is no "merging" of networks. Every client plays with the full, best-so-far network. The only thing that is "merged" is the training data, which consists of the network inputs, the search outcome and the game outcome sampled over many games. (The latter is described in the README "Training data format")

**Comment by wensdong** [comment]
Deepmind claims that they only trained for 3 days on 4 TPUs to reach a comparable level of the old Go. Nvadia claims their  Volta GPU probably have the better performance than the TPU. Anyone can explain the discrepancy why 1700years. That is a difference between normal PC and Quantum computer.

**Comment by ssj-gz** [comment]
@wensdong As far as I can tell, this is something that has been misreported by the media: when *playing* a game (e.g. against Alphago Lee), Alphago Zero does indeed use a single machine with 4 TPUs, but when generating the self-play games for training, it is practically certain that they use many more machines: generating 29 million self-play games on a single machine would be pretty tough :) As far as I'm aware, there's no data on the hardware used for self-play.

**Comment by gcp** [comment]
The idea is that all data stays open and public domain.

**Comment by gcp** [comment]
@wensdong See the post to computer-go that is linked from the README. It is simply not possible to get the required data with only that hardware, and nowhere did DeepMind claim that. They only talked about the playing machine (when the end result is already available), and the learning machine (64 GPU), which is a small part of the required learning pipeline.

**Comment by wensdong** [comment]
Yes,  4TPUs is only for inferencing. For the training, they said 4.9 million games self-played. 0.4 second per move probably is for a 4TPUs unit, which make the whole training(self-play) needs about 2000 such units. They also said they do not know the limit of AlphaGo zero, and they need resources to do something else.

**Comment by leehiufung911** [comment]
Is the server done yet? If so/when it's done I'd gladly let this train on my gtx1080 when I'm sleeping

**Comment by gcp** [comment]
Updates:

1) I managed to get good convergence (both on policy and value) on a small supervised network, and it's uploaded to the location linked in the README. Combined with a good GPU, it makes Leela Zero already play quite reasonably. I will use this to tune the search parameters (required for the learning procedure), and its strength will be a target to try to surpass reinforcement learning (self-learning).

2) I've rewritten my learning code to TensorFlow and open sourced it. With what's in the repo you can easily (if you have a working TensorFlow install...) learn a new supervised network, and the adaptation for the reinforcement learning with the server is trivial. The only thing missing is the automatic generation of symmetries to get more learning data (and maybe some optimization). Rewriting the code also flushed out a few bugs or missing details in the format specification.

3) Amazon has recently added NVIDIA Volta V100's to AWS, under the p3.* instances. Unlike TPU and the Google Cloud, they are available now, even to me, and they perform well for reasonable prices. I still don't plan to spend a lot of money renting thousands of instances, but they could come in very handy if we risk getting bottle-necked on the learning instead of the self-play.

With (2) now finished, it's time to do the final tweaks and start putting autogtp and leela-zero into a package that interacts with the server.

**Comment by wensdong** [comment]
I was thinking about how this distributed computing could work.  If it is then this "Shared-GPU" project could turn into a  business and undercut AWS. We can do the training in 1/10 of the price, while the GPU contributors can lease their GPU for credit.

**Comment by gcp** [comment]
Updates:

1) There's a feature-complete autogtp in the repository, but it'll need some small optimizations before being deployed more widely. Right now it downloads the network every game. The server needs an extra endpoint so autogtp can check if there's been an update without actually downloading it. Right now I'm using it for testing and seeing that we can get SGF and training data out of the db correctly.

2) Tuning for the search parameters is ongoing.

3) Given that the first 25k or so games are played by a random neural network, there's no point in actually using the GPU to play them out. I will probably hack up a Leela that just plays randomly and seed those myself, and start the distributed effort from epoch 2. (The thing the network learns from the first batch of games is who is won in a final position, i.e. how to count)

4) It's time to start estimating what kind of size network is feasible. I'm thinking of doing the first run with 6 residual layers of 128 filters, if it looks like that should progress at a fair speed (it's about 1/13th of the small Zero network, and 1/26th of the big Zero). This is about the size regular Leela uses when running on a CPU, and a fairly sweet spot for diminishing returns with supervised learning. Something like 1000 playouts? I'm not sure how the DeepMind guys selected 1600. Make this too big, and it's slow, make it too small, and the search doesn't have time to "discover" things the neural network doesn't know. It should at least be able to look at all the root moves, and explore the best for a few extra ply.

**Comment by marcocalignano** [comment]
I was looking at the code and I have a question:
why you need to process to play against each other? one process can receive the command 'genmove black' and after that 'genmove white'. 
Do you have a problem with the evaluation log for both players being in one game?

**Comment by gcp** [comment]
Good question! I need a controller program anyway to handle the HTTP networking, unzip etc and I already had some code to autoplay 2 GTP programs that I adapted. It is true that because the distributed program uses the same network for both players, it could play against itself in a single process. It's probably pretty easy to adapt autogtp to ditch the second process.

**Comment by marcocalignano** [comment]
I already did if you want to see the diff i can send it to you.
I would like to test it but there is no help on how to use autogtp

**Comment by gcp** [comment]
Sure, open a pull request and I'll take a look. 

autogtp needs to arguments, you can just launch it and it will connect to the testing server and do its thing.

**Comment by marcocalignano** [comment]
and which arguments?

**Comment by roy7** [comment]
Just run autogtp from the directory leelaz is in. It's magic.

**Comment by roy7** [comment]
He meant "no" arguments. :)

**Comment by marcocalignano** [comment]
It is stuck and this is the output 

autogtp v0.1
Best network hash: 
curl -s -C - -O -J -w %{filename_effective} http://zero-test.sjeng.org/best-network
Curl filename:

**Comment by gcp** [comment]
Can you open a separate issue to debug it?

**Comment by ssj-gz** [comment]
Note that you have to run it from within the "src" directory e.g.

cd leela-zero/src
../autogtp/autogtp

**Comment by roy7** [comment]
When limited to 400 playouts, it'll always do 400 new playouts each move right, even if prior search data is available from prior moves still being kept?

If so I wonder then if using 1 client to play both sides would give better results, as they'd share the same search space more often (since they value the board identically) and as each player adds playouts to the search tree/depth, that tree remains for the opposite player to immediately build upon.

**Comment by gcp** [comment]
Yes, they will be able to reuse each other's analysis from the transposition table. I guess it might reduce the randomness a bit, but there are sufficient sources of it that it shouldn't be an issue.

**Comment by marcocalignano** [comment]
I am working on an object Game so that the code in Main could be:

Game game(weightsname, &cerr);
    game.gameStart();
    do {
        game.move();
        if(!game.waitForMove())
            return false;
        game.readMove();
    }
    while(game.nextMove());
    if(game.getScore())
    {
        game.writeSgf();
        game.dumpTraining();
    }
    game.gameQuit();

This would allow for playing more games at the same time and then I had an Idea.
What if we give an Elo rating to each network. This Elo rating will be downloaded with the network. 
After playing n games with itself we can lat it learn from these games, then play against the original network and update the elo rating. Then push back the network with the rating and the server keeps the 10 best networks for the next download.

**Comment by gcp** [comment]
You need a good chunk of games for the training to progress, and the whole point of the effort is to gather these, learn, and distribute them to the clients as soon as a better one is found. Moreover, this would require each client to have a fully working TensorFlow install, which is definitely not what I'd want.

**Comment by marcocalignano** [comment]
I can understand your argument about the learning process. But why do you say it won't help anything running two game at the same time.
I did a little test and with one single game I did 4 games with a total 1818 move in 8min 56 sec so an average of 294 ms per move.
I did the let two game at the same time without the overhead of another autogpt process and i got 4 games with a total of 1460 moves in 4 min 48 sec so an average of 197 ms per move. Am I doing something wrong?

**Comment by gcp** [comment]
The point was that it helps now because the network you are fetching is a minuscule random one (1 residual block  x 8 filters, whereas even the smaller real networks are more like 6 blocks x 128 filters, and note the latter number causes a *quadratic* increase in computing time for every doubling), so the games' speed is limited by various overheads and not at all by the GPU.

If we put out the first trained network, the games will be much slower and peg the GPU close to 100%. So adding more processes is not expected to speed up things.

If you download the supervised network that is linked in the README.md, and hack up autogtp so it just uses that, I suspect you will see a negligible speedup from playing multiple games in parallel.

**Comment by gcp** [comment]
One thing that would be handy though - currently autogtp uses only the first GPU. I think it needs a commandline option to select the number of GPU so people with multiple GPU setups can use it (probably just forward the flags to the leelaz processes). That's one situation in which multiple processes would help.

**Comment by marcocalignano** [comment]
So my question is do I work on on the new Game object to be able to use different games on different GPUs? We could also think to use one on a CPU and one on a GPU.

**Comment by gcp** [comment]
If you have elegant ways to clean up the autogtp code I think that's never going to hurt, and yes, I think for the multiple GPU case it would be concretely useful.

**Comment by OmnipotentEntity** [comment]
Also requesting multiple GPU support.

Additionally, I'm still using hash d645af9 for self play, and it's the one I started with.  How often does this get updated?  You had mentioned 25k games of random play for the first epoch.  Is it 25k games per epoch?  How far along the second(?) one are we?

**Comment by gcp** [comment]
We are basically still testing the server and making sure the whole setup works fine.

I have found a small bug so far in the dumping of the learning data, and it occurred to me that the server must be able to signal to the client (i.e. autogtp) when it is out of date, so I will add this. I am also testing whether backing up and restoring the database works correctly. If I'm happy with this I'll make a release, put a note here, and change the URLs in autogtp to the final location. I understand @roy7 is writing some code to have a bit of a monitor page on the server so people can see the progress.

I will probably not pre-seed the database. Some people seem to have managed to build autogtp and connected to the test server with the result that we have almost 10k games now. By the time the above is fixed, we should have close to 25k so there's no point in bothering! (The games are arriving so quickly because the random network is very small, it will slow down substantially afterwards)

**Comment by roy7** [comment]
Nothing fancy yet but anyone can load http://zero-test.sjeng.org/ to see total number of games submitted so far.

**Comment by marcocalignano** [comment]
OK now I have TensorFlow installed where do I get a bunch of game to download so I can try to train the network?

**Comment by gcp** [comment]
https://u-go.net/gamerecords/ is a typical reference if you want to build a big DB for supervised learning.

Working on tooling to dump the contents of the server right now.

**Comment by bailsman** [comment]
Can I rent and donate an AWS server to this project? How much would it help?

**Comment by marcocalignano** [comment]
Can we also realize a way of Distribute learning? 
My idea was to divide the games in the database in chunks one for each of the N parties. 
1st step: Each remote party download one chunk, train the best network with that chunk and return the new network to the server. 
2nd step: Each party receive a new network that was previously trained by another(using other games) and train it with it own games and return it to the server.
...and so on
At the Nth step you have N networks that all have been trained with all data. At this point a tournament decide the best network
To even out the computation time every party that want to participate has to run a benchmark to establish how big will be his chunk of games so that all party end their learning process almost at the same time.

**Comment by gcp** [comment]
Distributing the learning over high latency, low bandwidth links will not work. Just training on separate chunks won't do anything useful as you have no good way to "average" the network. It requires a high bandwidth, low latency link so the errors from smaller batches can be averaged during the gradient calculation phase (which happens hundreds of times per second!). This is an area of very active research for all the biggest tech companies (Google, Microsoft, Facebook) whom are putting out papers in this area, as well as hardware (NVIDIA NVLink etc).

The solution you describe amounts to restarting the learning x times for a different random starting point. This can give a small improvement, but at the cost of occupying a bunch of machines that would have been more productively used to generate more data. If you can't resist doing this, it is probably smarter to simply run the learning with different meta-parameters (regularization, network size, optimization algorithm, learning rate) and see if you find an improvement.

There is no real point in distributing the learning. The bottleneck is generating the data, and I expect it to remain that way unless we get multiple thousands of clients...at which point reworking the TensorFlow code to support multiple GPU is probably much more useful and productive.

**Comment by Marcin1960** [comment]
Is there a link with a simple, idiot proof instruction how to donate GPU/CPU time for this project?

**Comment by gcp** [comment]
No, see issue #46. We need to do some final testing before this can really go live to a wider audience.

**Comment by gcp** [comment]
As promised here's a dump of the data from the first ~9300 games. You don't need to do anything with this, it's for people that want to fiddle.

http://sjeng.org/zero/train_2017_11_13.zip

I will probably dump every 25k games or so, and put it somewhere where it can be automatically indexed. I'll do the same for the SGF (haven't dumped those yet).

**Comment by marcocalignano** [comment]
What should I do with the new network afteri trained it?

**Comment by marcocalignano** [comment]
I was thinking: 
What if after i trained the network for a while I use this network to generate games with autogtp, will this improve the level of the games that I send to the server so the community network would gain a bit at least from my training?

**Comment by gcp** [comment]
Don't submit games with another network than the server sends you (in other words, don't modify autogtp). There are many problems with this, including not being able to tell if your client is broken or hostile, and you risk producing worse data if the server starts sending out an improved one that is better than yours.

To all who is reading here, please note that you do **not** need to any of the things I describe in the remainder of this post. It is an explanation of how the interaction with the server from my side works. You can reproduce it locally out of interest, but doing so isn't necessary and most likely will be of much less help than simply running the autogtp client, period.

(If you disregard all this and still train a network and it is **actually better**, we can upload it to the server. There is an API endpoint for this, which needs testing after which I'll put it here)

Now for testing whether it is better, the Zero paper uses 55% (+35 Elo) as a threshold. So you can play a match between the current network and your new network, and see if it wins by this margin in a statistically significant manner. (They use a fixed count of 400 games, which is not optimal. I'll get back to this). If it is, this is your new best network that should be sent to the clients.

If you try this now (I ran it overnight) it will fail because of an issue with how the engine behaves in games that are endlessly played out (which is very much the case with the random network). I'll file this in a minute or so.

I said previously that 400 games is not optimal, and this is where extending autogtp might come in useful. @marcocalignano, you noted previously that autogtp can be modified for doing this match, and I said that right now I just use gomill's ringmaster. This works fine for a fixed duration match, or a really obvious result (say new one wins 30-0). But there is a theoretical foundation for this kind of testing which could be implemented in autogtp itself called Sequential Probability Ratio Testing (SPRT). You basically give the tester two bounds, in our case this would be something like [25, 35], and 2 probabilities (typically 5%). The tester then runs the match until it decides that either the new version is less than 5% likely to lie in the interval. That is, it stops as soon as it is clear the new version is 35 Elo better, or less than 25 Elo better. You can find an implementation of SPRT is an autogtp like tool for chess, cutechess-cli: https://github.com/cutechess/cutechess/blob/master/projects/lib/src/sprt.h

If you want to do something with autogtp, then support for SPRT would be quite useful.

**Comment by marcocalignano** [comment]
Ok I will work on that then!

**Comment by marcocalignano** [comment]
can you give me a second file with other weights so I can test ?

**Comment by marcocalignano** [comment]
Or can I build one from the net dump that my training session did ?

**Comment by gcp** [comment]
Yes, it will write leelz-model.txt which is a valid weights file.

**Comment by marcocalignano** [comment]
OK it seems that playing against each other was quite easy to accomplish I see also that the trained network take a longer time per move.
Now I will start with the SPRT.

**Comment by gcp** [comment]
The default training parameters in tfprocess.py generate a much larger network than the one for the "random" weights that the server gives out now.

**Comment by gcp** [comment]
>So when you are ready to go live will you even do a windows binary?

Yes.

**Comment by gcp** [comment]
I think we're good. Updating the network and updating the client both caused an error, so those procedures are tested and fixed now 😉

I'll start the windows build and tag the version. Remember to update autogtp *and* the leelaz source, and replace the binary in your autogtp dir. There have been important fixes in both programs.

**Comment by marcocalignano** [comment]
Ok I have SPRT running on my Pc. I am testing the net I trained with the one that everyone download from your server. How can you check if my SPRT result is right?

**Comment by olbrichj** [comment]
I had autogtp running for a few games with the new network. seems to me that games are in general longer (500+ moves) than before (varying between 190 and 700+). I don't know what to expect with the changes. Just thought I'd report this.

**Comment by gcp** [comment]
One reason for the longer games on average is 7f4d8d25b0ad071d1486c54dc252f85a18dbd8b9.

There are occasionally also very short games now. I think it happens when due to the high randomness one engine "accidentally" passes, and then the other realizes passing a 2nd time wins (because it can count a bit now).

**Comment by olbrichj** [comment]
That explains it. Thanks!

Trying out the windows build. I don't really see an improvement compared to my Mac build. Is this running on the GPU?

**Comment by gcp** [comment]
The current network is very small so it's fairly CPU limited and mostly busy with various overheads. It'll remain like this for a bit while the data builds. No point in training a huge network with little data, it'll just overfit and be slower to generate data with.

**Comment by WhiteHalmos** [comment]
Haha, shortest game:
`1 (F18) 2 (L3) 3 (pass) 4 (pass) Game has ended.`

**Comment by olbrichj** [comment]
Is this data, which can be used for training? Or do we have to filter this out?

**Comment by marcocalignano** [comment]
I guess we need a game number limit. Just in case the network are close in value the Log Likelyhood Ratio will always oscillate around 0 never going near any of the bounds.

**Comment by gcp** [comment]
>Is this data, which can be used for training? Or do we have to filter this out?

Why would it be filtered out? The program needs to learn how to count.

**Comment by sbbdms** [comment]
Is this version (0.4) ready to go live? Or is it still a test version?
I am not sure whether I should describe in a notification.

**Comment by sbbdms** [comment]
> There are occasionally also very short games now. I think it happens when due to the high randomness one engine "accidentally" passes, and then the other realizes passing a 2nd time wins (because it can count a bit now).

Can Leela-zero really "count" when an "accidentally pass" (especially by black in an early stage of a game) happens?
There's a very short game mentioned in this issue above:

> 1 (F18) 2 (L3) 3 (pass) 4 (pass) Game has ended.
Score: W+7.5

I also notice that there are some very short games when I am running autogtp:

![w 7 5 84](https://user-images.githubusercontent.com/19167292/32878264-42a16c16-cae0-11e7-836c-216a82673146.png)
![w 7 5 72](https://user-images.githubusercontent.com/19167292/32878269-47470bfe-cae0-11e7-80ec-3522f1beeab9.png)
![w 7 5 8](https://user-images.githubusercontent.com/19167292/32878270-49ad96f6-cae0-11e7-991a-f4f3f0c233c2.png)

All their results are W+7.5.
They probably mean Leela-zero cannot make clear estimations of these games. The results are just concluded from the komi.
Will these submitted games do benefit/harm to the network?

**Comment by gcp** [comment]
>Can Leela-zero really "count" when an "accidentally pass" (especially by black in an early stage of a game) happens?

I think it must have some inkling. Otherwise it's very unlikely to get 2 passes in a row, and this is obviously happen fairly regularly now.

>Will these submitted games do benefit/harm to the network?

The program will learn that black cannot pass early, which as you can clearly see, is important knowledge.

**Comment by gcp** [comment]
1) As the comment you quote says, this is near the sweet spot for supervised learning (network size about 12-14 x 128-192), and thus it is also a good point of reference to see where the program ends up in strength. Reducing the number of filters seems to have fairly fast diminishing returns. So making it less than 64-128 does not really help so much (there are factors here like the efficiency of the GPU computation and being able to keep it busy), and making it larger than 256-384 has not really demonstrated much benefit (doubling cause a quadratic increase in computation time, so the cost is high).

2) I don't think it is possible to know for sure unless you run that experiment. The Zen authors made the interesting observation that for a 20 block network, the "Master" program (using Monte Carlo playouts, supervised learning start, human features) was actually stronger than Zero using 20 blocks. But 40 blocks Zero was (much!) stronger again. So did they not mention a 40 block Master because the approach failed (the program getting stuck in an optimum and not progressing), or did they not mention it because making a point about it learning by itself was more important than getting the strongest program? It is clear though, that learning from human weights leads to a trained network that is far from the optimal. So in a sense, one is taking a risk to make it harder (or not possible) to drag it away from that wrong setting to a more optimal one, in return for saving some computation time.

**Comment by a22063821** [comment]
Have you considered BOINC?

**Comment by olbrichj** [comment]
Just a question for my understanding. Currently the distributed process looks like this:  
- Server distributes the current best network
- Client loads the best network
- Client playes game with best network against best network
- Client uploads game and result
- Repeat till an amount of games has been accumulated
- Server (or developer) trains new network with (all? or only the new?) games
- Server tests if new network is above 55% winrate (I read this in one of the other tickets, maybe I understood it wrong and how many games are played?)
- Server distributes new network in case it's better
- Repeat all

Is this correct?

What I'm wondering is, whether this server tests could also be distributed e.g. distribute the current best network and the newly trained network to clients. Get the results and when you have a statistically relevant amount, check if it's better and act accordingly.

I'm not so deep in NN, so I'm also wondering, if you could use these strength testing games, as input for your next network.. or does it have to be the same version?

How would this influence the result?

In case this would be possible, wouldn't it make sense for the server to act a little bit smarter, distribute 2 networks (maybe twice the same) and train a new network every time a version is (dis-)proven?

**Comment by marcocalignano** [comment]
That could be a good Idea. The training is done on the server side but the generation of games and verification of the networks could be done in a distributed way.
autogtp could switch between single network game mode to competition mode depending on the difference  (or absence of thereof) of the downloaded networks.
We would use better the distributed resources. and would be a faster verification.

**Comment by gcp** [comment]
Self-play games for learning are played with different parameters as the test for the strongest one - they use more randomization to prevent repeat games and to give some probability the search to discover moves that the policy network thinks are not good.

The amount of games to test for whether a new network is an improvement is minuscule compared to those required for learning, so whether these are added to the dataset or not is largely irrelevant. But note the whole point of "guarding" the learning by the check whether the new network is actually better is exactly to prevent starting to play self-play games with something that is weaker.

The training is done over a sliding window of the last 500 000 games.

The only way testing the new networks would be a bottleneck is if the actual self-play games (which ARE the bottleneck, and remember they will come in much slower when the network is larger...) would start coming in very, very quickly, we first parallelize the trainer for multiple GPU so it can also keep up with the game inflow (as this part can't be distributed, see above), and the differences between the old and new networks are such that an SPRT test would take a large amount of games to terminate. I'm all for planning a bit ahead but that's a large pile of if's.

**Comment by WhiteHalmos** [comment]
Definitely a good idea to avoid premature optimization. Right now generating self-play games seems like the vastly bigger opportunity to optimize for, i.e. using multiple GPUs, more efficient use of existing resources, recruiting more help, etc.

For reference, @gcp do you mind sharing some timing for training a new network next time you train one? I'm assuming it's on the order of < 1 hour on your machine/server?

**Comment by gcp** [comment]
If one is looking for ways to speed up the project, most useful IMHO would be:

a) BOINC client (from what I can see, would need quite a bit of development, including changes to the engine so the clients take a seed from the server and behave more deterministic, but it looks like it might be possible)

b) cuDNN support for NVIDIA cards, combined with batching the neural net evaluation, or any equivalent for AMD cards.

But neither of those things are easy, obviously.

**Comment by gcp** [comment]
>For reference, @gcp do you mind sharing some timing for training a new network next time you train one? I'm assuming it's on the order of < 1 hour on your machine/server?

The training can run forever. Because the current dataset (24k-ish games) is relatively tiny, it mostly stops improving after an hour or so but I let the training run while I start the game test. I have another machine that trains a smaller network, to see if a smaller network isn't better due to less overfitting.

**Comment by marcocalignano** [comment]
> Self-play games for learning are played with different parameters as the test for the strongest one 
@gcp can you give me these parameters so I can test them too (btw I am almost finish with the SPRT support)

**Comment by marcocalignano** [comment]
So How would the command line of leelaz look like?

**Comment by marcocalignano** [comment]
leelaz -g -q -r ?

**Comment by gcp** [comment]
/leelaz --gtp --noponder -p 1000 -r 0 -w 9k.txt

**Comment by marcocalignano** [comment]
@gcp You can get the new autogtp code from my fork repository in the branch sprt. 
The command line looks like that
./autogtp -c -n firstnet -n secondnet 
I am still testing (played 44 games and still Likelyhood Ratio 1.38181 Lower Bound -2.94444 Upper Bound 2.94444) maybe you can test it too and tell if you find something

**Comment by gcp** [comment]
@marcocalignano It's looking good. But to be practically useful right now, it should be possible to play games in parallel. I think you'll want this capability for multi-GPU support too.

**Comment by marcocalignano** [comment]
Ok I can do that but the question is how many games at the same time?

**Comment by gcp** [comment]
@marcocalignano configurable? I play 16 at the same time now (with engine set to 1 thread).

For multi-GPU, you can assume that the GPUs in the --gpu options are in the correct order (i.e. if the users asks for 2 GPU, use --gpu 0 and --gpu 1 to get the cards that are most likely correct. I will fix leelaz to ensure that is right).

**Comment by marcocalignano** [comment]
@gcp OK in the same branch on my fork you find the last version of autogtp the command line is:

Usage: ./autogtp [options]
autogtp

Options:
  -h, --help                Displays this help.
  -v, --version             Displays version information.
  -c, --competition         Play two networks  against each other.
  -g, --gamesNum <num>      Play 'gamesNum' games on one GPU at the same time.
  -u, --gpus <num>          Index of the GPU for multiple GPUs support
  -n, --network <filename>  Networks to use ad players (two are needed).

./autogtp -c -g 16  -n  net1 -n net2     -> single gpu total of 16 games 
./autogtp -c -g 8 -u 0 -u 1  -n  net1 -n net2     -> two GPUs 8 games per GPU total of 16 games

I do not have a second GPU and my is really slow, so the program is still running and I didn't manage to test it till the end. Maybe you can test it on you (faster) machine.
The version is 1000 on purpose so it won't interfere with the version for the games generation if someone think of compile this code.

On the other end I could bring the option -g and -u in the 'normal' autogtp to generate more games, since many people ask for it.

**Comment by marcocalignano** [comment]
@gcp you can leave it for now I found a bug! (after 12 h :( )

**Comment by marcocalignano** [comment]
@gcp Now it seems to work better. Remember on my PC it takes a long time so there could be still some glitches. 
I will have to add a stat summary at the end but you can still test it.
If you say it is OK I will apply the same setting for the game generation part.

**Comment by marcocalignano** [comment]
@gcp did you have time to check it out I am trying to use these option to implement multi games generation

**Comment by gcp** [comment]
@marcocalignano I will test it tomorrow, I think by then I will have some 85k networks to test with it too.

**Comment by marcocalignano** [comment]
But I should go on with the multi games option? 
I also try to rebase on the next from your repository

**Comment by gcp** [comment]
Yes, this looks good. Specifying the GPU id explicitly works but requires the user to understand how to launch Leela on the console and find the right numbers (no problem for my use of course). I think for using this technique in autogtp it should take a simple count of # of GPU to use, and then just use --gpu 0, --gpu 1 etc. As said, I will change leela-zero so it numbers the GPUs in order of preference.

**Comment by marcocalignano** [comment]
Ok I am ready to test multi games mode what do I have to puul to have the last client?

**Comment by marcocalignano** [comment]
Game data 852c24aaf81ef708dc773157e298c0a296d6c9e4ab9ba8d9ceddaeb0686ded50 stored in database

Game data 661291685827292130b0d566cd76cb47c65a593fd4c8469bee3e495de565cf06 stored in database


Can you check these two game if they are ok?

**Comment by marcocalignano** [comment]
@gcp did you check the previous games?

**Comment by gcp** [comment]
They arrived fine in the db, i.e.:

```
"(;GM[1]FF[4]RU[Chinese]DT[2017-11-19]SZ[19]KM[7.5]PB[Leela Zero 0.3]PW[Human]RE[B+11.5]\n\n;B[oa];W[md];B[ch];W[dk];B[mf];W[nh];B[sc];W[mg];B[di];W[jf]\n;B[le];W[mi];B[ob];W[qo];B[gb];W[fc];B[lb];W[eq];B[sh];W[ne]\n;B[rm];W[bg];B[ao];W[pr];B[hs];W[el];B[ek];W[pj];B[gn];W[ab]\n;B[ib];W[pq];B[jh];W[bd];B[fn];W[ka];B[la];W[fg];B[hh];W[ba]\n;B[ci];W[dl];B[qi];W[ca];B[pl];W[nl];B[eb];W[ai];B[lj];W[hp]\n;B[rn];W[em];B[pf];W[lh];B[jp];W[fh];B[ii];W[bf];B[dc];W[db]\n;B[jm];W[hj];B[dd];W[of];B[ll];W[gg];B[lf];W[he];B[on];W[om]\n;B[nf];W[fk];B[lr];W[mb];B[oq];W[ed];B[fe];W[en];B[ki];W[ld]\n;B[fp];W[fo];B[dg];W[ee];B[ef];W[aa];B[qg];W[il];B[eh];W[af]\n;B[lm];W[cj];B[gr];W[gi];B[bj];W[me];B[im];W[kj];B[cc];W[bc]\n;B[sr];W[lq];B[jj];W[po];B[nc];W[ic];B[qs];W[ck];B[ss];W[as]\n;B[ep];W[sg];B[ec];W[kk];B[fi];W[hm];B[pm];W[io];B[gc];W[cs]\n;B[bi];W[gd];B[se];W[gm];B[fs];W[fj];B[lg];W[hd];B[kq];W[mr]\n;B[ke];W[ge];B[ha];W[dm];B[jl];W[jg];B[cf];W[gj];B[oi];W[oh]\n;B[nd];W[ql];B[ds];W[rl];B[je];W[fd];B[rg];W[bm];B[rc];W[cg]\n;B[sd];W[eg];B[sq];W[hc];B[pn];W[df];B[hi];W[ni];B[ml];W[ik]\n;B[nk];W[cd];B[ri];W[hf];B[gh];W[fb];B[bk];W[if];B[kc];W[co]\n;B[fl];W[re];B[og];W[lc];B[qh];W[jo];B[dr];W[gp];B[hg];W[go]\n;B[kg];W[lk];B[nn];W[ei];B[id];W[dh];B[gl];W[eh];B[ej];W[dq]\n;B[kh];W[jk];B[oe];W[hq];B[oo];W[am];B[mo];W[dj];B[cm];W[dg]\n;B[cl];W[kb];B[li];W[tt];B[cp];W[ph];B[mc];W[ra];B[kp];W[rf]\n;B[qp];W[jc];B[oj];W[eo];B[ji];W[ps];B[cn];W[jb];B[ip];W[mm]\n;B[sf];W[mj];B[gk];W[od];B[ej];W[cq];B[kr];W[dn];B[ag];W[pd]\n;B[nm];W[qk];B[kd];W[ek];B[mq];W[lc];B[qf];W[ea];B[js];W[md]\n;B[ir];W[ma];B[rj];W[hn];B[br];W[gf];B[ld];W[me];B[ja];W[fm]\n;B[bn];W[lp];B[pk];W[kf];B[ia];W[hk];B[op];W[mh];B[mk];W[hl]\n;B[bq];W[mp];B[ls];W[qd];B[gk];W[rd];B[is];W[bl];B[pc];W[es]\n;B[qa];W[de];B[ne];W[jn];B[km];W[kn];B[bp];W[md];B[gq];W[er]\n;B[lb];W[al];B[rr];W[la];B[pe];W[pg];B[fr];W[lo];B[ns];W[lc]\n;B[mn];W[rq];B[qm];W[ro];B[nq];W[fa];B[ol];W[ce];B[np];W[ho]\n;B[bs];W[ij];B[ac];W[dp];B[mm];W[jr];B[ng];W[fq];B[fp];W[aj]\n;B[pa];W[rk];B[qr];W[do];B[an];W[ej];B[pi];W[cr];B[iq];W[ln]\n;B[ak];W[cf];B[ms];W[bm];B[hr];W[qe];B[jq];W[sp];B[so];W[hb]\n;B[qj];W[in];B[bh];W[fl];B[si];W[fn];B[of];W[gl];B[oc];W[ap]\n;B[am];W[fi];B[nb];W[ig];B[ad];W[ga];B[om];W[gc];B[ko];W[ja]\n;B[ib];W[gb];B[ks];W[sa];B[nr];W[gn];B[nj];W[ph];B[bl];W[mj]\n;B[qc];W[rd];B[pg];W[sj];B[oh];W[qd];B[aq];W[rf];B[ph];W[bb]\n;B[mi];W[cb];B[lh];W[pd];B[cc];W[ie];B[rp];W[ha];B[sk];W[pp]\n;B[ni];W[jd];B[jr];W[gk];B[ap];W[ff];B[kl];W[da];B[mj];W[lb]\n;B[qe];W[mh];B[qq];W[ef];B[rs];W[ia];B[ae];W[ib];B[nh];W[ah]\n;B[ds];W[dc];B[mr];W[cc];B[qn];W[be];B[sl];W[ae];B[qk];W[dd]\n;B[sj];W[ad];B[rk];W[pb];B[ql];W[ec];B[tt];W[ac];B[rl];W[rb]\n;B[no];W[os];B[tt];W[dr];B[nl];W[id];B[mg];W[ag];B[rh];W[ep]\n;B[or];W[qo];B[bo];W[pr];B[pp];W[os];B[mh];W[ih];B[tt];W[po]\n;B[me];W[qb];B[tt];W[sn];B[al];W[fp];B[od];W[tt];B[tt])
```

**Comment by marcocalignano** [comment]
Ok I did them with  the multigames version. With all the improvement merge that comes in  ;)  I have to keep rebasing but I'd like to do a pull request maybe on the next branch

**Comment by kimitsu** [comment]
Question: can measures be implemented to prevent simple attacks on Leela? Like checking some statistical characteristics of the input data per user? My only fear is that later on or even right now someone is spamming training data with noise.

**Comment by WhiteHalmos** [comment]
Currently since Leela Zero uses two threads, the result isn't deterministic. So confirming results isn't really possible.

**Comment by OmnipotentEntity** [comment]
You can give each thread its own random engine and deterministically split work, so it's definitely possible.  However, you'd need to implement your own distribution functions in order to make it replicable across platforms.

It would also require server support, because the server would need to select a random engine initialization sequence for each thread and each run.

I have quite a bit of experience in making reproducible pseudorandom results across platforms with multithreaded execution, and if you need I can take a look at the code and see what I can do.

**Comment by gcp** [comment]
The problem isn't the random number generators, the problem is that the threads are used to feed hardware that is used by other processes.

**Comment by OmnipotentEntity** [comment]
I'm not certain that work stealing algorithms necessarily preclude reproducibility.  This is might just be something that no one's tried, rather than a breakthrough.

You're right that the approach in my previous post is completely wrong.  (Because this is a tree search, whereas my previous work had been on a 2d plane.)

However, we can shift the problem like this: (assuming we need x bits of random numbers maximum per node visit)  We don't know which nodes we will visit, so we need a way of numbering nodes, if we come up with a way, we can simply bin a bunch of random numbers and pull from the proper area of the array.  There's a rub though, and that's coming up with a method of numbering nodes such that we don't need to bin gigabytes of RNG, due to how explosively huge the search space is.

If it is possible to order nodes by their desired visit order deterministically (ie, like an A* search), then the scheme will work.  Otherwise a small and simple hashing scheme might be doable (think 16-24 bit), but probably would ultimately screw up the randomness and likely lead to worse training data due to hash collisions reusing RNG data, (birthday paradox says about 1% collisions at 16-bit (that's about 10 per move), and 6e-3% for 24-bit (that's about 10 per game), but this requires generating 64*x kb and 16*x Mb of RNG engine calls per move respectively).

Another possible scheme is some sort of (cryptographic?) hash function that takes a naive numbering scheme (such as breadth first numbering, would require gmp which can be expensive) and a single random number from the engine and uses the hash as the RNG bits (if x is more than the hash length, it's possible to generate more hashes by appending sequential bytes to the end of the message).  But again, this has some question marks surrounding just how good of random numbers you'll get (but this is at least probably fine.)  

Honestly, I should actually read the paper so I know what the hell I'm talking about.  So if I'm that far off base, just tell me to RtFP.  And I totally understand if all this complexity isn't really warranted, as the probability of bad actors is rather low.

**Comment by gcp** [comment]
I don't want to ruin your fun in exploring that topic, but as for the problem at hand, it's just easier to run 2 instances with 1 thread, if we worry about bad actors. 

The multi-threading is important for playing real games with a clock - where running a 2nd instance does nothing useful -  but for just generating the data it is not needed. After removing multi-threading, I suspect making the engine reproducible is not that difficult, although the floating point accuracy of GPUs (how stable are those?) may be an issue.

Requiring independently verified results cuts the throughput in half, but it seems to be how BOINC works. I think that may be required if a larger, more long term run, on the order of the full Alpha Go Zero network size, would be done.

**Comment by OmnipotentEntity** [comment]
It's not just GPUs.  CPUs also have floating point reproducibility issues.  For instance, the RSQRTPS instruction is only required to give an estimate, and on Haswell CPUs (at least) the lower 12 bits of the mantissa are always all zeros:

`rsqrt(1.162357688 (3F94C823)) = precise 0.927534580 (3F6D72E8) | rsqrtps 0.927490234 (3F6D7000)
rsqrt(1.745346546 (3FDF6784)) = precise 0.756936014 (3F41C68F) | rsqrtps 0.756835938 (3F41C000)
rsqrt(1.880780697 (3FF0BD6C)) = precise 0.729173601 (3F3AAB1F) | rsqrtps 0.729248047 (3F3AB000)
rsqrt(1.839437246 (3FEB72AE)) = precise 0.737322569 (3F3CC12C) | rsqrtps 0.737304688 (3F3CC000)
rsqrt(1.323228836 (3FA95F90)) = precise 0.869325697 (3F5E8C21) | rsqrtps 0.869506836 (3F5E9800)
rsqrt(1.288511395 (3FA4EDF1)) = precise 0.880959332 (3F61868D) | rsqrtps 0.880981445 (3F618800)
rsqrt(1.943412185 (3FF8C1BB)) = precise 0.717327595 (3F37A2C8) | rsqrtps 0.717285156 (3F37A000)
rsqrt(1.910214782 (3FF481EB)) = precise 0.723533928 (3F393985) | rsqrtps 0.723510742 (3F393800)
rsqrt(1.700642347 (3FD9AEA6)) = precise 0.766820133 (3F444E53) | rsqrtps 0.766845703 (3F445000)
rsqrt(1.856116772 (3FED953C)) = precise 0.734002173 (3F3BE791) | rsqrtps 0.734008789 (3F3BE800)`

**Comment by hot-bit-git** [comment]
Hi, do you still consider launching BOINC project for Leela?
